### PR TITLE
Power Law Fit Function for Muon Modelling

### LINKED ISSUE #32665: Power Law Fit Function
**Is your feature request related to a problem? Please describe.**
Users would like to use the Power Law as a fit function for fitting Muon data. Formula is as follows:-
```y = a x^b +c```

**Describe the solution you'd like**
Add a Power Law fit function to Mantid and update the Fit Function Tree to make it easy to find for Muon users.

Additional context
Part of a series of new fitting functions for Muon results analysis - #32578

### PR DESCRIPTION
**Pull request by sf1919**
**Description of work.**
Added a new Power Law Fit function for Muon modelling and associated documentation.

Also added links to Gaussian, Lorenztian and Polynomial Fitting functions under MuonModelling of the fitting function tree as part of the WIMDA project.

**To test:**

TBC

Fixes #32665 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/CurveFitting/inc/MantidCurveFitting/Functions/PowerLaw.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+</UL>
+*/
+
+class MANTID_CURVEFITTING_DLL PowerLaw : public BackgroundFunction {
+public:
+  /// overwrite IFunction base class methods
+  std::string name() const override { return "PowerLaw"; }
+  void function1D(double *out, const double *xValues, const size_t nData) const override;
+  void functionDeriv1D(API::Jacobian *out, const double *xValues, const size_t nData) override;
+  const std::string category() const override { return "Muon\\MuonModelling"; }
```
Comment (AnthonyLim23) [inline Framework/CurveFitting/inc/MantidCurveFitting/Functions/PowerLaw.h:34]: might be worth asking in tech QA if anyone else would like this? I can see it being in "general"

File: Framework/CurveFitting/src/Functions/PowerLaw.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+using namespace CurveFitting;
+
+using namespace Kernel;
+
+using namespace API;
+
+DECLARE_FUNCTION(PowerLaw)
+
+void PowerLaw::init() {
```
Comment (AnthonyLim23) [inline Framework/CurveFitting/src/Functions/PowerLaw.cpp:23]: can we have some better defaults?

File: Framework/CurveFitting/src/Functions/PowerLaw.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  declareParameter("a", 0.0, "coefficient for linear term");
+  declareParameter("b", 0.0, "exponent");
+  declareParameter("c", 0.0, "coefficient for constant term");
+}
+
+void PowerLaw::function1D(double *out, const double *xValues, const size_t nData) const {
+  const double a = getParameter("a");
```
Comment (AnthonyLim23) [inline Framework/CurveFitting/src/Functions/PowerLaw.cpp:30]: can we have some better parameter names (e.g. magnitude, exponent, constant)?

File: Framework/CurveFitting/src/Functions/PowerLaw.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+void PowerLaw::functionDeriv1D(Jacobian *out, const double *xValues, const size_t nData) {
+  const double a = getParameter("a");
+  const double b = getParameter("b");
+  const double c = getParameter("c");
+
+  for (size_t i = 0; i < nData; i++) {
+    double diffa = pow(xValues[i], b);
+    double diffb = a * b * pow(xValues[i], b - 1) * log(xValues[i]);
```
Comment (AnthonyLim23) [inline Framework/CurveFitting/src/Functions/PowerLaw.cpp:46]: The exponent is `b` and not `b-1`. We could also call the `function1D` and the multiply it be the log(x)

File: docs/source/fitting/fitfunctions/PowerLaw.rst [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+================
+PowerLaw
+================
+
+.. index:: PowerLaw
+
+Description
+-----------
+
+A power law background function is defined as:
```
Comment (AnthonyLim23) [inline docs/source/fitting/fitfunctions/PowerLaw.rst:12]: lets say something like: "A power law may be used to describe <insert experiment data>"

File: docs/source/release/v6.3.0/muon.rst [hunk 6]
```diff
@@ -9,4 +9,15 @@ MuSR Changes
     putting new features at the top of the section, followed by
     improvements, followed by bug fixes.
 
+Muon Analysis and Frequency Domain Analysis
+-------------------------------------------
+
+New Features
+############
+- Added a :ref:`Power Law <func-PowerLaw>` function to MuonModelling Fit Functions.
```
Comment (AnthonyLim23) [inline docs/source/release/v6.3.0/muon.rst:17]: also worth putting in the framework (see my note above about adding it to general)

**Comment by AnthonyLim23** [comment]
there is an error in the derivative if you use the following script, changing between numerical and analytic derivatives gives very different answers. 

```

x =np.linspace(5, 10,200)

def func(x, a,b,c):
    return a +b*(np.power(x,c))

def calc_error(ran):
    return 1. + 0.*(ran-0.5)

def error(f):
    return [y*calc_error(np.random.random()) for y in f]
    return 

constant = 2.5
mag = 1.2
expon = 3.56

param=[mag, expon, constant]

y = error(func(x,constant,mag,expon))
ws = CreateWorkspace(x,y)

f = PowerLaw()#Magnitude=1, Constant=0, Exponent=1)
c = CompositeFunction(Gaussian(PeakCentre=1))
myFunc = CompositeFunction(f, NumDeriv=True)


fit_output = Fit(InputWorkspace=ws, Output='fit', Function=myFunc)
paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters

print("The fit was: {}".format(fit_output.OutputStatus))
for j in range(len(paramTable)-1):
    print(paramTable.column(0)[j], "is", paramTable.column(1)[j]," error is", 100*(paramTable.column(1)[j]-param[j])/param[j])


```

**Review by AnthonyLim23 — DISMISSED** [review]
Tried a few different data sets and the fits converge nicely. For one case the analytic derivatives led to a better set of parameters.

**Review by AnthonyLim23 — APPROVED** [review]

**Review by DavidFair — APPROVED** [review]
LGTM lets :shipit:

**Pull request merged**